lib/src/view/analysis/tree_view.dart

```diff
@@ -162,124 +131,696 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  final IList<PgnComment>? rootComments;
+
+  final _PgnTreeViewParams params;
+
+  @override
+  State<_PgnTreeView> createState() => _PgnTreeViewState();
+}
+
+typedef _Subtree = ({
+  _MainLinePart mainLinePart,
+
+  /// This is nullable since the very last mainline part might not have any sidelines.
+  _IndentedSideLines? sidelines,
+  bool containsCurrentMove,
+});
+
+class _PgnTreeViewState extends State<_PgnTreeView> {
+  /// Caches the result of [_mainlineParts] to avoid recalculating it on every rebuild.
+  List<List<ViewNode>> mainlineParts = [];
+
+  /// Caches the top-level subtrees, where each subtree is a [_MainLinePart] and its sidelines.
+  /// Building the whole tree is expensive, so we cache the subtrees that did not change when the current move changes.
+  List<_Subtree> subtrees = [];
+
+  UciPath _mainlinePartOfCurrentPath() {
+    var path = UciPath.empty;
+    for (final node in widget.root.mainline) {
+      if (!widget.params.pathToCurrentMove.contains(path + node.id)) {
+        break;
+      }
+      path = path + node.id;
+    }
+    return path;
+  }
+
+  void _rebuildChangedSubtrees({required bool fullRebuild}) {
+    var path = UciPath.empty;
+    subtrees = mainlineParts.mapIndexed(
+      (i, mainlineNodes) {
+        final mainlineInitialPath = path;
+
+        final sidelineInitialPath = UciPath.join(
+          path,
+          UciPath.fromIds(
+            mainlineNodes
+                .take(mainlineNodes.length - 1)
+                .map((n) => n.children.first.id),
+          ),
+        );
+
+        path = sidelineInitialPath;
+        if (mainlineNodes.last.children.isNotEmpty) {
+          path = path + mainlineNodes.last.children.first.id;
+        }
+
+        final mainlinePartOfCurrentPath = _mainlinePartOfCurrentPath();
+        final containsCurrentMove =
+            mainlinePartOfCurrentPath.size > mainlineInitialPath.size &&
+                mainlinePartOfCurrentPath.size <= path.size;
+
+        if (fullRebuild ||
+            subtrees[i].containsCurrentMove ||
+            containsCurrentMove) {
+          // Skip the first node which is the continuation of the mainline
+          final sidelineNodes = mainlineNodes.last.children.skip(1);
+          return (
+            mainLinePart: _MainLinePart(
+              params: widget.params,
+              initialPath: mainlineInitialPath,
+              nodes: mainlineNodes,
+            ),
+            sidelines: sidelineNodes.isNotEmpty
+                ? _IndentedSideLines(
+                    sidelineNodes,
+                    parent: mainlineNodes.last,
+                    params: widget.params,
+                    initialPath: sidelineInitialPath,
+                    nesting: 1,
+                  )
+                : null,
+            containsCurrentMove: containsCurrentMove,
+          );
+        } else {
+          // Avoid expensive rebuilds by caching parts of the tree that did not change across a path change
```

Comment:
One might get confused here: by "expensive rebuilds" you mean actually the framework's `State.build`, but this comment is inside a method named `_rebuildChangedSubtrees`. I think it is worth the precision in the comment.